interface/src/avatar/OtherAvatar.cpp

```diff
@@ -56,7 +56,7 @@ OtherAvatar::~OtherAvatar() {
 
 void OtherAvatar::removeOrb() {
     if (!_otherAvatarOrbMeshPlaceholderID.isNull()) {
-        DependencyManager::get<EntityScriptingInterface>()->deleteEntity(_otherAvatarOrbMeshPlaceholderID);
+        DependencyManager::get<EntityScriptingInterface>()->deleteEntity(_otherAvatarOrbMeshPlaceholderID, true);
```

Comment:
I think there must be something deeper happening here.  Can we find the commit that introduced the crash and determine what changed in the logic?

Reply:
Yes, almost certainly. The problem is that I don't understand how things work here well enough.

My initial thought was that since `entity->die()` just sets a flag, this must be some sort of sanity check -- don't delete entities that are still present in the domain, or something along those lines. And if you create an entity in the interface, you need to mark it as dead yourself before deleting it.

The problem is that the full intention of the code isn't very clear. 

`die()`/`isDead()` are defined in `SpatiallyNestable.h`. and do nothing besides setting a boolean. Being dead gets checked at a few points: `isReadyToComputeShape`, `shouldBePhysical`, `OtherAvatar::shouldBeInPhysicsSimulation`. So that part seems pretty clear: something dead is something that's about to be cleaned up, so it's not physical and shouldn't be rendered. Considering there are threads involved, cleanup may not be immediate, and meanwhile we don't want people bumping into stuff that's gone.

This code was added back in 2015 in 6eb177091bd and apparently the change there was moving those functions from `AvatarData.h`. Then both asserts and several entity->die() were added to `EntitySimulation.h`. 

That's the point at which it gets confusing for me. The above seems to be just dealing with a delayed cleanup. But in that case just removing an entity outright isn't really a problem. So what is it that the assert and the `entity->die()` in `Entitysimulation.h` are trying to achieve?

Reply:
there must have been a more recent change in the merge with master that is triggering this.  I wonder if it could be specific to local entities?  they should never be physical, so maybe they used to be handled differently.

Reply:
I'm looking into it, yes.

I'm just trying to figure out what's the intent here -- all this die/isDead business would seem to serve some purpose beyond simply getting stuff out of the way when it's scheduled for cleanup.

So far it looks to me like this flag has two purposes: getting removed stuff out of the way, and a second one that performs some sort of correctness checking.

Reply:
I think the crash after this PR is applied is interesting: The backtrace comes from QtScript, which directly calls `EntityScriptingInterface::deleteEntity`. Which I take means some script is deleting something outright, and there's no opportunity there for `entity->die()` to be called.

There's a change in commit 4159bc48 that looks like it could be relevant, I'm looking into it.

Reply:
Another promising looking commit: 9d92546a

Reply:
Okay, pushed an update. There was a single `theEntity->die()` removed in commit 9d92546. I added it back and that seems to fix things.

Reply:
By the way, for testing you need to run a server. This problems can happen on both sides.